Add str() support to LineMatcher

mcsitter opened the pull request:
Added `__str__` method to LineMatcher class.

The `LineMatcher` class is used in the `RunResult` class, which has the `stdout` attribute, which is an instance of `LineMatcher`. `result.stdout.str()` is used 66 times in the codebase. The use of the `str()` function is mentioned [here](https://github.com/pytest-dev/pytest/blob/8ea8cdb36d031206742e609975ba0393533db92c/src/_pytest/pytester.py#L512-L517) as well. If I should make any other changes. Since `.str()` is used throughout the codebase, changing everything to `str(matcher)` seems pointless.

If you think a test should be added, let me now, however it is implicitly tested in the `Pytester` tests.

- [X] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. 

Closes #1265

bluetech commented on src/_pytest/pytester.py:
```diff
@@ -1707,6 +1707,10 @@ def __init__(self, lines: List[str]) -> None:
         self._log_output: List[str] = []
 
+    def __str__(self) -> str:
+        """Return the entire original text."""
```
```suggestion
        """Return the entire original text.
        
        .. versionadded:: 6.2
            You can use :method:`str` in older versions.
        """
```

bluetech reviewed (commented):
Thanks @mcsitter, I left some comments.

We should also update this sentence to refer to `str(...)` instead of `.str()` IMO: https://docs.pytest.org/en/latest/reference.html#pytest.pytester.RunResult.stdout

mcsitter commented on src/_pytest/pytester.py:
```diff
@@ -1908,4 +1912,4 @@ def _fail(self, msg: str) -> None:
     def str(self) -> str:
         """Return the entire original text."""
-        return "\n".join(self.lines)
+        return self.__str__()
```
Makes sense🤦‍♂️

mcsitter commented:
Thanks for the feedback! Not much to discuss about the requested changes.

I will try to formulate better changelog entries in the future and make myself more familiar with sphinx directives. Is there a sensible guideline for when to use them? There seem to be so many, that it could clutter up things quite a bit.

What would you put instead of ```:func:`stdout.str() <LineMatcher.str()>` ```? I figure that is the part that needs to change.
```:func:`str(stdout) <LineMatcher.__str__()>` ```? Would that reference work?

bluetech commented on src/_pytest/pytester.py:
```diff
@@ -1707,6 +1707,14 @@ def __init__(self, lines: List[str]) -> None:
 
+    def __str__(self) -> str:
+        """Return the entire original text.
+        
```
Oops, my suggestion added trailing whitespace here that should be removed:
```suggestion

```

bluetech commented:
> Is there a sensible guideline for when to use them? There seem to be so many, that it could clutter up things quite a bit.

I like hyperlinks so I tend to add them when someone might be inclined to follow them. But there are not strict guidelines, just depends on personal preference and laziness :)

> What would you put instead of :func:`stdout.str() <LineMatcher.str()>` ? I figure that is the part that needs to change.
:func:`str(stdout) <LineMatcher.__str__()>` ? Would that reference work?

Right, I think that's the correct incantation, but only sphinx can tell :)

BTW in the Checks section of every PR there is a `docs/readthedocs.org:pytest` link to the rendered docs, so you can check that the reference works even without testing locally if you prefer.

nicoddemus commented:
> If you think a test should be added, let me now, however it is implicitly tested in the Pytester tests.

I really think a test is needed, mostly to ensure we don't regress this in the future by accident.

mcsitter commented:
1. I updated the reference. Have a look at it [here](https://pytest--8050.org.readthedocs.build/en/8050/reference.html#pytest.pytester.RunResult.stdout).

2. I added a test.

nicoddemus reviewed (approved):
LGTM!
